File: Java/TirePressureMonitoringSystem/src/test/java/tddmicroexercises/tirepressuremonitoringsystem/TestAlarm.java
package tddmicroexercises.tirepressuremonitoringsystem;


import org.junit.*;
import org.mockito.Mockito;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

public class TestAlarm {

    private final Sensor sensor = mock(Sensor.class);

    @Test
    public void withinBound() {
        doTest(20.0, false);
    }

    private void doTest(double value, boolean result) {
        when(sensor.popNextPressurePsiValue()).thenReturn(value);
        Alarm alarm = new Alarm(sensor);
        alarm.check();
        assertThat(alarm.isAlarmOn(), is(result));
    }

    @Test
    public void onUpperBound() {
        doTest(21.0, false);
    }

    @Test
    public void onLowerBound() {
        doTest(17.0, false);
    }

    @Test
    public void overUpperBound() {
        doTest(21.1, true);
    }

    @Test
    public void underLowerBound() {
        doTest(16.9, true);
    }
}

File: kotlin/TirePressureMonitoringSystem/src/test/kotlin/tddmicroexercises/tirepressuremonitoringsystem/TestAlarm.kt
package tddmicroexercises.tirepressuremonitoringsystem

import org.hamcrest.CoreMatchers
import org.hamcrest.CoreMatchers.*
import org.junit.Test
import org.junit.Assert.*
import org.mockito.Mockito.*

class TestAlarm {

    @Test fun withinBounds() = doTest(20.0, false)
    @Test fun upperBound() = doTest(21.0, false)
    @Test fun lowerBound() = doTest(17.0, false)
    @Test fun overUpperBound() = doTest(21.1, true)
    @Test fun underLowerBound() = doTest(16.9, true)

    val sensor: Sensor = mock(Sensor::class.java)

    private fun doTest(value: Double, result: Boolean) {
        `when`(sensor.popNextPressurePsiValue()).thenReturn(value)
        val alarm = Alarm(sensor)
        alarm.check()
        assertThat(alarm.isAlarmOn, `is`(result))
    }
}
